File: packages/icoder-python/icoder_sdk/pagination.py
"""Lazy, bounded cursor pagination helpers."""

from __future__ import annotations

from collections.abc import Callable, Iterable, Iterator
from typing import Generic, Optional, TypeVar


PageT = TypeVar("PageT")
ItemT = TypeVar("ItemT")


class CursorPager(Generic[PageT, ItemT]):
    """Iterate cursor pages lazily and fail closed on cursor loops."""

    def __init__(
        self,
        fetch_page: Callable[[Optional[str]], PageT],
        items: Callable[[PageT], Iterable[ItemT]],
        next_page_token: Callable[[PageT], Optional[str]],
        *,
        initial_page_token: Optional[str] = None,
        max_pages: int = 10000,
    ) -> None:
        if not isinstance(max_pages, int) or isinstance(max_pages, bool) or max_pages < 1:
            raise ValueError("max_pages must be a positive integer")
        self._fetch_page = fetch_page
        self._items = items
        self._next_page_token = next_page_token
        self._initial_page_token = initial_page_token
        self._max_pages = max_pages

    def pages(self) -> Iterator[PageT]:
        token = self._initial_page_token
        seen = {token} if token else set()
        for _ in range(self._max_pages):
            page = self._fetch_page(token)
            yield page
            raw_next_token = self._next_page_token(page)
            if raw_next_token is None or raw_next_token == "":
                return
            if not isinstance(raw_next_token, str):
                raise RuntimeError("iCoDer pagination returned an invalid page token")
            next_token = raw_next_token
            if next_token in seen:
                raise RuntimeError("iCoDer pagination returned a repeated page token")
            seen.add(next_token)
            token = next_token
        raise RuntimeError(f"iCoDer pagination exceeded max_pages={self._max_pages}")

    def __iter__(self) -> Iterator[ItemT]:
        for page in self.pages():
            yield from self._items(page)


class PageNumberPager(Generic[PageT, ItemT]):
    """Lazily iterate page-number results using an authoritative total count."""

    def __init__(
        self,
        fetch_page: Callable[[int], PageT],
        items: Callable[[PageT], Iterable[ItemT]],
        total_items: Callable[[PageT], int],
        *,
        initial_page: int = 1,
        max_pages: int = 10000,
    ) -> None:
        if not isinstance(initial_page, int) or isinstance(initial_page, bool) or initial_page < 1:
            raise ValueError("initial_page must be a positive integer")
        if not isinstance(max_pages, int) or isinstance(max_pages, bool) or max_pages < 1:
            raise ValueError("max_pages must be a positive integer")
        self._fetch_page = fetch_page
        self._items = items
        self._total_items = total_items
        self._initial_page = initial_page
        self._max_pages = max_pages

    def pages(self) -> Iterator[PageT]:
        emitted = 0
        for index in range(self._max_pages):
            page = self._fetch_page(self._initial_page + index)
            page_items = tuple(self._items(page))
            total = self._total_items(page)
            if not isinstance(total, int) or isinstance(total, bool) or total < 0:
                raise RuntimeError("iCoDer pagination returned an invalid total")
            yield page
            emitted += len(page_items)
            if not page_items or emitted >= total:
                return
        raise RuntimeError(f"iCoDer pagination exceeded max_pages={self._max_pages}")

    def __iter__(self) -> Iterator[ItemT]:
        for page in self.pages():
            yield from self._items(page)

File: packages/icoder-sdk/src/pagination.ts
export interface CursorPageAdapter<TPage, TItem> {
  items(page: TPage): readonly TItem[];
  nextPageToken(page: TPage): string | null | undefined;
}

export interface CursorPagerOptions {
  initialPageToken?: string;
  maxPages?: number;
}

export interface PageNumberPagerOptions {
  initialPage?: number;
  maxPages?: number;
}

export interface PageNumberAdapter<TPage, TItem> {
  items(page: TPage): readonly TItem[];
  totalItems(page: TPage): number;
}

/** Bounded cursor pager with repeated-token detection and lazy requests. */
export class AsyncCursorPager<TPage, TItem> implements AsyncIterable<TItem> {
  private readonly maxPages: number;

  constructor(
    private readonly fetchPage: (pageToken?: string) => Promise<TPage>,
    private readonly adapter: CursorPageAdapter<TPage, TItem>,
    private readonly options: CursorPagerOptions = {},
  ) {
    this.maxPages = options.maxPages ?? 10000;
    if (!Number.isInteger(this.maxPages) || this.maxPages < 1) {
      throw new RangeError('maxPages must be a positive integer');
    }
  }

  async *pages(): AsyncGenerator<TPage> {
    let token = this.options.initialPageToken;
    const seen = new Set<string>();
    if (token) seen.add(token);
    for (let pageIndex = 0; pageIndex < this.maxPages; pageIndex += 1) {
      const page = await this.fetchPage(token);
      yield page;
      const rawNext = this.adapter.nextPageToken(page);
      if (rawNext === null || rawNext === undefined || rawNext === '') return;
      if (typeof rawNext !== 'string') {
        throw new Error('iCoDer pagination returned an invalid page token');
      }
      const next = rawNext;
      if (seen.has(next)) throw new Error('iCoDer pagination returned a repeated page token');
      seen.add(next);
      token = next;
    }
    throw new Error(`iCoDer pagination exceeded maxPages=${this.maxPages}`);
  }

  async *[Symbol.asyncIterator](): AsyncGenerator<TItem> {
    for await (const page of this.pages()) {
      for (const item of this.adapter.items(page)) yield item;
    }
  }
}

/** Lazy page-number pager that terminates from the authoritative total count. */
export class AsyncPageNumberPager<TPage, TItem> implements AsyncIterable<TItem> {
  private readonly initialPage: number;
  private readonly maxPages: number;

  constructor(
    private readonly fetchPage: (page: number) => Promise<TPage>,
    private readonly adapter: PageNumberAdapter<TPage, TItem>,
    options: PageNumberPagerOptions = {},
  ) {
    this.initialPage = options.initialPage ?? 1;
    this.maxPages = options.maxPages ?? 10000;
    if (!Number.isInteger(this.initialPage) || this.initialPage < 1) {
      throw new RangeError('initialPage must be a positive integer');
    }
    if (!Number.isInteger(this.maxPages) || this.maxPages < 1) {
      throw new RangeError('maxPages must be a positive integer');
    }
  }

  async *pages(): AsyncGenerator<TPage> {
    let emitted = 0;
    for (let index = 0; index < this.maxPages; index += 1) {
      const page = await this.fetchPage(this.initialPage + index);
      const items = this.adapter.items(page);
      const total = this.adapter.totalItems(page);
      if (!Number.isInteger(total) || total < 0) {
        throw new Error('iCoDer pagination returned an invalid total');
      }
      yield page;
      emitted += items.length;
      if (items.length === 0 || emitted >= total) return;
    }
    throw new Error(`iCoDer pagination exceeded maxPages=${this.maxPages}`);
  }

  async *[Symbol.asyncIterator](): AsyncGenerator<TItem> {
    for await (const page of this.pages()) {
      for (const item of this.adapter.items(page)) yield item;
    }
  }
}
